File: game/js/skills.js
/* =========================================================
 *  ハグミン ツムツム  ―  スキル効果
 *  api: {
 *    tsums(), field:{w,h}, clear(list, tag), addTime(sec),
 *    makeBig(n), flash(color), shake(power), ring(x,y,r,color)
 *  }
 * =======================================================*/

var Skills = (function () {

  function pickRandom(arr, n) {
    var a = arr.slice(), out = [];
    while (a.length && out.length < n) out.push(a.splice(Math.floor(Math.random() * a.length), 1)[0]);
    return out;
  }

  var TABLE = {

    /* まんまるバースト：円形範囲を消す */
    burst: function (api, power) {
      var ts = api.tsums();
      if (!ts.length) return;
      // ツムが密集している場所を中心にする
      var cx = 0, cy = 0;
      ts.forEach(function (t) { cx += t.x; cy += t.y; });
      cx /= ts.length; cy /= ts.length;
      cy = Math.max(cy, api.field.h * 0.55);
      api.ring(cx, cy, power, '#FFD27F');
      api.shake(10);
      var hit = ts.filter(function (t) {
        return (t.x - cx) * (t.x - cx) + (t.y - cy) * (t.y - cy) <= power * power;
      });
      api.clear(hit, 'skill');
    },

    /* フレイムライン：よこ帯 */
    line_h: function (api, power) {
      var ts = api.tsums();
      if (!ts.length) return;
      var ys = ts.map(function (t) { return t.y; }).sort(function (a, b) { return a - b; });
      var cy = ys[Math.floor(ys.length * 0.55)];
      api.ring(api.field.w / 2, cy, power * 1.4, '#FF7F6E');
      api.band(0, cy - power / 2, api.field.w, power, '#FF7F6E');
      api.shake(9);
      api.clear(ts.filter(function (t) { return Math.abs(t.y - cy) <= power / 2; }), 'skill');
    },

    /* そらとびウェーブ：たて帯 */
    line_v: function (api, power) {
      var ts = api.tsums();
      if (!ts.length) return;
      var cx = 0;
      ts.forEach(function (t) { cx += t.x; });
      cx /= ts.length;
      cx = Math.max(power / 2, Math.min(api.field.w - power / 2, cx + (Math.random() - 0.5) * 120));
      api.band(cx - power / 2, 0, power, api.field.h, '#7FC9F2');
      api.shake(9);
      api.clear(ts.filter(function (t) { return Math.abs(t.x - cx) <= power / 2; }), 'skill');
    },

    /* めばえの祝福：大ツムを作る */
    grow: function (api, power) {
      var n = api.makeBig(power);
      api.flash('rgba(157,224,160,0.45)');
      api.shake(5);
      return n;
    },

    /* ハートクロス：じゅうじに消す */
    cross: function (api, power) {
      var ts = api.tsums();
      if (!ts.length) return;
      var cx = 0, cy = 0;
      ts.forEach(function (t) { cx += t.x; cy += t.y; });
      cx /= ts.length; cy /= ts.length;
      cy = Math.max(cy, api.field.h * 0.58);
      api.band(0, cy - power / 2, api.field.w, power, '#FFA9CF');
      api.band(cx - power / 2, 0, power, api.field.h, '#FFA9CF');
      api.shake(10);
      api.clear(ts.filter(function (t) {
        return Math.abs(t.y - cy) <= power / 2 || Math.abs(t.x - cx) <= power / 2;
      }), 'skill');
    },

    /* よぞらのねがい：ランダム消し＋時間 */
    night: function (api, power) {
      var ts = api.tsums();
      var hit = pickRandom(ts, power);
      hit.forEach(function (t) { api.ring(t.x, t.y, t.r * 1.6, '#B6A6F5'); });
      api.addTime(3);
      api.flash('rgba(182,166,245,0.4)');
      api.shake(7);
      api.clear(hit, 'skill');
    },

    /* ハグ・オールクリア：同じ種類を全部消す */
    same: function (api, power) {
      var ts = api.tsums();
      if (!ts.length) return;
      var counts = {};
      ts.forEach(function (t) { counts[t.charId] = (counts[t.charId] || 0) + 1; });
      var ids = Object.keys(counts).sort(function (a, b) { return counts[b] - counts[a]; });
      var targets = ids.slice(0, power);
      api.flash('rgba(255,216,77,0.5)');
      api.shake(12);
      api.clear(ts.filter(function (t) { return targets.indexOf(t.charId) >= 0; }), 'skill');
    }
  };

  function run(skillId, api, power) {
    var fn = TABLE[skillId];
    if (fn) fn(api, power);
  }

  return { run: run };
})();

File: tsumtsum/engine/hagukumin/skills.py
"""キャラクターのスキル効果.

各関数は (消すツムのリスト, 見せかたの指示, 追加の時間) を返す。
見せかたの指示（effects）はクライアントがそのまま演出に使う。
"""

from __future__ import annotations

from typing import Callable, Dict, List, Tuple

from .board import Board, WIDTH
from .physics import Body

Result = Tuple[List[Body], List[Dict], float]


def _center_of(bodies: List[Body]) -> Tuple[float, float]:
    cx = sum(b.x for b in bodies) / len(bodies)
    cy = sum(b.y for b in bodies) / len(bodies)
    return cx, cy


def burst(board: Board, power: float, color: str) -> Result:
    """まんまるに まとめて消す。"""
    tsums = board.tsums()
    if not tsums:
        return [], [], 0.0
    cx, cy = _center_of(tsums)
    cy = max(cy, board.height * 0.55)
    hit = [b for b in tsums if (b.x - cx) ** 2 + (b.y - cy) ** 2 <= power * power]
    fx = [{"fx": "ring", "x": round(cx, 1), "y": round(cy, 1),
           "r": round(power, 1), "color": color},
          {"fx": "shake", "power": 10}]
    return hit, fx, 0.0


def rainbow(board: Board, power: float, color: str) -> Result:
    """よこ一列を にじで 消す。"""
    tsums = board.tsums()
    if not tsums:
        return [], [], 0.0
    ys = sorted(b.y for b in tsums)
    cy = ys[int(len(ys) * 0.55)]
    half = power / 2.0
    hit = [b for b in tsums if abs(b.y - cy) <= half]
    fx = [{"fx": "band", "x": 0, "y": round(cy - half, 1),
           "w": WIDTH, "h": round(power, 1), "color": color, "rainbow": True},
          {"fx": "shake", "power": 9}]
    return hit, fx, 0.0


def wave(board: Board, power: float, color: str) -> Result:
    """たて一列を なみで 消す。"""
    tsums = board.tsums()
    if not tsums:
        return [], [], 0.0
    cx, _ = _center_of(tsums)
    half = power / 2.0
    cx = max(half, min(WIDTH - half, cx + (board.rng.random() - 0.5) * 120))
    hit = [b for b in tsums if abs(b.x - cx) <= half]
    fx = [{"fx": "band", "x": round(cx - half, 1), "y": 0,
           "w": round(power, 1), "h": board.height, "color": color},
          {"fx": "shake", "power": 9}]
    return hit, fx, 0.0


def cross(board: Board, power: float, color: str) -> Result:
    """じゅうじに 消す。"""
    tsums = board.tsums()
    if not tsums:
        return [], [], 0.0
    cx, cy = _center_of(tsums)
    cy = max(cy, board.height * 0.58)
    half = power / 2.0
    hit = [b for b in tsums
           if abs(b.y - cy) <= half or abs(b.x - cx) <= half]
    fx = [{"fx": "band", "x": 0, "y": round(cy - half, 1),
           "w": WIDTH, "h": round(power, 1), "color": color},
          {"fx": "band", "x": round(cx - half, 1), "y": 0,
           "w": round(power, 1), "h": board.height, "color": color},
          {"fx": "shake", "power": 10}]
    return hit, fx, 0.0


def ramune(board: Board, power: float, color: str) -> Result:
    """ランダムに消して 時間を足す。"""
    tsums = board.tsums()
    n = int(power)
    board.rng.shuffle(tsums)
    hit = tsums[:n]
    fx = [{"fx": "ring", "x": round(b.x, 1), "y": round(b.y, 1),
           "r": round(b.r * 1.6, 1), "color": color} for b in hit]
    fx.append({"fx": "flash", "color": color})
    fx.append({"fx": "shake", "power": 7})
    return hit, fx, 3.0


def dam(board: Board, power: float, color: str) -> Result:
    """ツムを 大ツムに そだてる（消さない）。"""
    grown = board.make_big(int(power))
    fx = [{"fx": "ring", "x": round(b.x, 1), "y": round(b.y, 1),
           "r": round(b.r * 1.6, 1), "color": color} for b in grown]
    fx.append({"fx": "flash", "color": color})
    return [], fx, 0.0


TABLE: Dict[str, Callable[[Board, float, str], Result]] = {
    "burst": burst,
    "rainbow": rainbow,
    "wave": wave,
    "cross": cross,
    "ramune": ramune,
    "dam": dam,
}


def run(skill_id: str, board: Board, power: float, color: str) -> Result:
    fn = TABLE.get(skill_id)
    if fn is None:
        raise KeyError(f"unknown skill: {skill_id}")
    return fn(board, power, color)
